File: my-calendar-i/my-calendar-i.py
class MyCalendar:

    def __init__(self):
        self.li = []

    def book(self, start: int, end: int) -> bool:
        i = 0
        # print(self.li)
        if not self.li:
            self.li.append([start, end])
            return True
        while i < len(self.li):
            if start >= self.li[i][1]:
                j = i + 1
                if j >= len(self.li) or end <= self.li[j][0]:
                    self.li.insert(j, [start, end])
                    return True
            elif end <= self.li[i][0]:
                j = i - 1
                if j < 0 or start >= self.li[j][1]:
                    self.li.insert(i, [start, end])
                    return True
            i += 1
        return False
        
        
        


# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(start,end)

File: my-calendar-i/my-calendar-i.c

struct bst{
    struct bst* left, *right;
    int start;
    int end;
};

bool insert(struct bst* root, struct bst* node){
    if (root->end <= node->start){
        if(!root->right){
            root->right = node;
            return true;
        }
        return insert(root->right, node);
    }else if (root->start >= node->end){
        if(!root->left){
            root->left = node;
            return true;
        }
        return insert(root->left, node);
    }
    return false;
}

void freebst(struct bst* node){
    if(!node)
        return;
    freebst(node->left);
    freebst(node->right);
    free(node);
}

struct bst* newbst(int start, int end){
    struct bst* node = (struct bst*)malloc(sizeof(struct bst));
    memset(node, 0, sizeof(struct bst));
    node->start = start;
    node->end = end;
    return node;
}

typedef struct {
    struct bst* root;
} MyCalendar;


MyCalendar* myCalendarCreate() {
    MyCalendar* ret = malloc(sizeof(MyCalendar));
    ret->root = NULL;
    return ret;
}

bool myCalendarBook(MyCalendar* obj, int start, int end) {
    struct bst* node;
    bool ret;
    if(!obj->root){
        obj->root = newbst(start, end);
        return obj->root;
    }
    node = newbst(start, end);
    ret = insert(obj->root, node);
    if(!ret){
        free(node);
    }
    return ret;
}

void myCalendarFree(MyCalendar* obj) {
    freebst(obj->root);
    free(obj);
}

/**
 * Your MyCalendar struct will be instantiated and called as such:
 * MyCalendar* obj = myCalendarCreate();
 * bool param_1 = myCalendarBook(obj, start, end);
 
 * myCalendarFree(obj);
*/